### ISSUE TITLE
Use asymmetric GPG encryption

### ISSUE THREAD

**Issue by piegamesde** [duplicati/duplicati#3074]
I don't like that one of the most precious things I own, the encryption password of my backup, is stored in plain text somewhere on my hard drive. I'd like to have the option to be asked for the password every time Duplicati wants to access that backup. I am using KeePass password manager with browser integration so I don't mind having to "type" it often.

An advanced implementation could also make use of native password safes like Gnome-Keyring etc. or writing a custom KeePass plugin.

**Comment by Pectojin** [comment]
The keychain idea is also described in #2024

The problem with prompting for the encryption key is that Duplicati can't do anything without the encryption password. So that means no automatic backups and no automatic verification.

**Comment by piegamesde** [comment]
Yes, automatic features won't work with this, but for me it would be worth it.

Backups can still run in a scheduled way without the password. They just require asking for it every time. A little push notification "Duplicati needs the database password to continue" would be enough. But I'm not asking for that much.

**Comment by wjansenw** [comment]
You don't need to store the pwd if you use gpg encryption. You do need the private key if you want to compact or restore but otherwise you don't need it.

**Comment by piegamesde** [comment]
Are there tutorials/documentation out there how to use GPG encryption and is it possible to change encryption on existing backups? I'll try it out tomorrow

**Comment by wjansenw** [comment]
You'll need to have gpg installed to start with. Generate a private/public
key for encryption. The private key is only needed for decryption, so you
could store it somewhere else (even print it (eg. as a qr code)). You
can/should also pwd protect the gpg key. If you want to compact, duplicati
needs access to the key. But you could perfectly do without, and compact
manually from time to time. To restore, access to the key is also needed.

I suggest you first try it out with a test backup.

The default settings in Duplicati when using gpg are symmetric encryption
which you don't want, as you will still need to provide duplicati with the
pwd at every run. You will need to set one of the advanced gpg options to
specify "--encryption" to change this default setting. I'll export my
configuration later so you can have a look.

Yes, it is possible to change existing backup. However you will need to
download all files, run a python script to decrypt and reencrypt (available
on GitHub since a week or two), and reupload all files. Once done, delete
(or change) your configuration to point to the new files and encryption
settings. Let Duplicati rebuild the db.  Possible but for advanced users
only...

I succesfully did it on a small backup set.

Regards,
Wim

**Comment by piegamesde** [comment]
So I can't get it up to work. The python script outputs an error:

```
~/Downloads » python3 ReEncrypt.py -c config.txt                                                                                        piegames@piec
decrypting: .../duplicati-20180205T211433Z.dlist.zip.aes to /tmp/tmp466mt5py
encrypting: /tmp/tmp466mt5py .../duplicati-20180205T211433Z.dlist.zip.gpg
Traceback (most recent call last):
  File "ReEncrypt.py", line 197, in <module>
    main(sys.argv[1:])
  File "ReEncrypt.py", line 193, in main
    mainReEncrypt(options)
  File "ReEncrypt.py", line 42, in mainReEncrypt
    encrypt(options['new'],temp_file.name,options['new']['passwd'], dlist_reenc_fullpath)
  File "ReEncrypt.py", line 114, in encrypt
    status  = gpg.encrypt_file(f, recipients=options['recipients'], output=encrypted, armor=False)
AttributeError: 'GPG' object has no attribute 'encrypt_file'
```

**Comment by wjansenw** [comment]
I followed the examples in this post:
https://www.saltycrane.com/blog/2011/10/python-gnupg-gpg-example/
http://pythonhosted.org/python-gnupg/

Can you verify the version of python-gnupg?

**Comment by piegamesde** [comment]
Sorry, my bad, I missed the "recipients" tag in the config file. What does it represent, what should I put in there and why?

And you said that by default the gpg encryption is still symmetric. How can I tell the script if I want symmetric/asymmetric gpg encryption?

**Comment by wjansenw** [comment]
The python script encrypts with public key so is not symmetric. 

In duplicati you need to set the option --encrypt at gpg encrypt options.

**Comment by wjansenw** [comment]
Recipients are the users that need to be able to decrypt. In this case the user you entered while generating the private key.

**Comment by piegamesde** [comment]
I successfully re-encrypted my whole backup. Then I changed the Duplicati settings to the new location and the encryption method.

> you need to set the option --encrypt at gpg encrypt options

I assume you mean the `gpg-encryption-switches` under "Options" in the backup settings and set it to `--encrypt` even if it says: "Use this switch to specify any extra options to GPG. You cannot specify the --passphrase-fd option here. The --encrypt option is always specified."

I then verified the database and obviously it failed telling me to repair the local database. I am now doing this (it asked me for my GPG password so it seems to be working).

The problem is: the operation is tearing my backup apart, deleting it file by file. From the 7530 files in the original backup folder (200GB) only 1400 are left (60MB) and constantly shrinking. This is interesting because rebuilding the local database should not modify anything on the remote. Here's how the log looks like:

```
Mar 7, 2018 3:05 PM: Backend event: Delete - Started: duplicati-i7c8e87f53c89497c87b360a0da104275.dindex.zip.gpg (91.40 KB)
Mar 7, 2018 3:05 PM: Failed to accept new index file: duplicati-i7c8e87f53c89497c87b360a0da104275.dindex.zip.gpg, message: Volume duplicati-be78071c828d74fab8f5605139e05e613.dblock.zip.gpg has local state Deleting
{"ClassName":"System.Exception","Message":"Volume duplicati-be78071c828d74fab8f5605139e05e613.dblock.zip.gpg has local state Deleting","Data":null,"InnerException":null,"HelpURL":null,"StackTraceString":"  at Duplicati.Library.Main.Operation.RepairHandler.RunRepairRemote () [0x004ff] in <118ad25945a24a3991f7b65e7a45ea1e>:0 ","RemoteStackTraceString":null,"RemoteStackIndex":0,"ExceptionMethod":null,"HResult":-2146233088,"Source":"Duplicati.Library.Main"}
Mar 7, 2018 3:05 PM: Backend event: Get - Completed: duplicati-i7c8e87f53c89497c87b360a0da104275.dindex.zip.gpg (91.40 KB)
Mar 7, 2018 3:05 PM: Backend event: Get - Started: duplicati-i7c8e87f53c89497c87b360a0da104275.dindex.zip.gpg ()
Mar 7, 2018 3:05 PM: Backend event: Delete - Completed: duplicati-i7c8b3d0d854746ee8f8dbd2b0f560e48.dindex.zip.gpg (122.28 KB)
Mar 7, 2018 3:05 PM: Backend event: Delete - Started: duplicati-i7c8b3d0d854746ee8f8dbd2b0f560e48.dindex.zip.gpg (122.28 KB)
```

Have I messed something up along the way or is this a bug?

**Comment by wjansenw** [comment]
I have the following options set:
```
--gpg-encryption-command=--encrypt
--gpg-decryption-command=--decrypt
--gpg-encryption-switches=--recipient my_email@mail.com
```
Did you do a repair or a delete/repair? I think a repair will not work, and it should have been a delete/repair of the DB.

**Comment by piegamesde** [comment]
Do I have to set these options in the `gpg-encryption-switches` or somewhere else?

Why should a delete/repair work and not a repair? And why does a local repair destroy my remote backup?

By the way: All files are deleted now, and Duplicati like:
```
{"ClassName":"Duplicati.Library.Interface.UserInformationException","Message":"Repair not possible, missing 792 blocks.\nIf you want to continue working with the database, you can use the \"list-broken-files\" and \"purge-broken-files\" commands to purge the missing data from the database and the remote storage.","Data":null,"InnerException":null,"HelpURL":null,"StackTraceString":"  at Duplicati.Library.Main.Operation.RepairHandler.RunRepairRemote () [0x00e53] in <118ad25945a24a3991f7b65e7a45ea1e>:0 ","RemoteStackTraceString":null,"RemoteStackIndex":0,"ExceptionMethod":null,"HResult":-2146233088,"Source":"Duplicati.Library.Main"}
```
Wow, Sherlock, congratulations!

**Comment by wjansenw** [comment]
No, it are 3 different advanced options. You can add them in the advanced options section at the end of the backup config.

I have never tried a repair when changing encryption, but as the files are so drastically changed, basically all information in the db is incorrect. Extensions changed, hashsums changed, .... I would go for a delete & recreate of the db. Or even better: delete your old config, and create a new one.

**Comment by piegamesde** [comment]
Thank you a lot for your help, it worked fine on my little test data set. A few questions before I try it again on my big backup:

 - Does the password I enter for the GPG encryption matter? If yes, what is it used for? If no, why can't I leave it blank?
 - If it does not matter, why does changing it still break things and requires a database rebuild?
 - Is the GPG key password cached somewhere? I did a few actions (even restarting Duplicati), but only had to enter it once.

**Comment by wjansenw** [comment]
For (public-key) encryption, password is not needed. However, I think Duplicati makes it a mandatory parameter, but it is not used in the case of public-key encryption. I see in the code it is being passed to the gpg-command, but I guess that only makes sense if you use GPG in a symmetric encryption set-up.

For restoration or compacting, where decryption is needed, the private key needs to be avalable & unlocked. From my tests so far, this is only possible from the commandline, and I had to enter the pwd to a GPG-popup that was provided. So, also in this case, I believe the pwd is not used. But I need to test this a bit further.

Changing the pwd of a private key has no impact on your encrypted results. But I believe Duplicati's GUI will not allow you to change the pwd (or it at least gives you a warning). 

If it is not clear, please find some info on GPG public key encryption. You better know what you are doing, otherwise you will end up with a backup that can't be restored...

**Comment by piegamesde** [comment]
Thank you a lot for your help. I created a simple Wiki page with the steps I've done for future users. (Maybe split into one for re-encrypting and one for asymetric GPG in general?). Please have a look at it because of mistakes.

Since the script is quite slow and the write speed was not the bottleneck for me, I changed the script to run several files in parallel. I'll push this one soon.

**Comment by wjansenw** [comment]
I'll be happy to review. Just point me to the link!

**Comment by piegamesde** [comment]
https://github.com/duplicati/duplicati/wiki/Re-encrypt-database

**Comment by Pectojin** [comment]
The other issue is more focused on hiding the plain text passwords in the frontend, so it may be best to keep a separate issue for the keychain feature.

Perhaps change the title of this issue to reflect a feature request for keychain integration? 

Or alternatively make a new issue to track discussion/work on keychain implementation. That would keep this issue pure from those discussions and I think the discussion here was quite good as it describes quite well the use case for PGP over AES and how to change to PGP.
